sfdx-cicd.sh

```diff
@@ -1,54 +1,70 @@
 #!/bin/bash
 set -o errexit -o pipefail -o noclobber -o nounset
 
-# Download sfdx binary and setup
-wget -q $SFDX_DOWNLOAD_URL
-mkdir ~/sfdx-cli
-tar xJf sfdx-linux-x64.tar.xz -C ~/sfdx-cli --strip-components 1
-export PATH=~/sfdx-cli/bin:$PATH
+# Install sfpowerkit (only if missing)
+if [[ -z $(sf plugins | grep sfpowerkit) ]]; then
+  echo 'y' | sf plugins install sfpowerkit
+fi
+
+# Set logging level
+SF_LOG_LEVEL='fatal'
 
-# Install sfpowerkit
-echo 'y' | sfdx plugins:install sfpowerkit
+# Auth sfdx into org with auth url
+# Generate this with:
+#  sf org display --verbose --json (pass in sfdxAuthUrl value starting with force:// )
+echo $SFDX_AUTH_URL > sfdx_auth.txt
+sf org login sfdx-url --sfdx-url-file sfdx_auth.txt --set-default --alias SFOrg
+rm sfdx_auth.txt
 
+echo "sfdx diff command: sf sfpowerkit project diff -r $REVISION_FROM -t $REVISION_TO -d diffdeploy"
+# Prepare diff
+sf sfpowerkit project diff -r $REVISION_FROM -t $REVISION_TO -d diffdeploy
 
+# Set Validate_only flag
 VALIDATE_ONLY=`echo $VALIDATE_ONLY | tr '[:upper:]' '[:lower:]'`
 if [[ $VALIDATE_ONLY = true ]]; then
-  VALIDATE_FLAG='-c';
+  VALIDATE_FLAG='--dry-run'
 elif [[ $VALIDATE_ONLY = false ]]; then
-  VALIDATE_FLAG='';
+  VALIDATE_FLAG=''
 else
   echo "Bad Validate param...choose true or false"
   exit 2
 fi
 
+# Set Test_Level (or use org default)
 TEST_LEVEL=`echo $TEST_LEVEL | tr '[:upper:]' '[:lower:]'`
+
+# Override sandbox default (notestrun) if TEST_LEVEL not specified
+# but package contains test class changes
+if [[ -n $(find diffdeploy/ -name '*test.cls') ]] && [[ $TEST_LEVEL = '' ]]; then
```

Comment:
I think this is trying to be a little too clever, and also isn't correct.

First, with this approach, you'd have to use `-iname` for it to stand a chance. Second, plenty of test classes in plenty of orgs don't end in the word `test`. Our own org has quite a few. Sometimes the start with test, sometimes they end with it, sometimes they omit it all together.

If you want to go down that route, you really need to be grepping for `@istest` case insensitive within the body of the file. Even that might not be 100% in the case of comments or other edge cases I haven't thought about, but it will avoid issues with either people not following a convention or not having a convention.

The problem though this only runs the tests when the body of the test has changed. What you really want to do is test code that has changed by test classes that cover it, as historically test cases do not change very often. This is the problem we started to solve at RH but they didn't allow us to invest enough time to solve it (happy to talk about the solution over chat).

In short, I would just pull this out and make people be explicit about it vs trying to assume what they meant for them.